pkg/clusters/addons/loadimage/addon.go

```diff
@@ -0,0 +1,65 @@
+package loadimage
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters"
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters/types/kind"
+)
+
+// -----------------------------------------------------------------------------
+// CertManager Addon
+// -----------------------------------------------------------------------------
+
+const (
+	// AddonName indicates the unique name of this addon.
+	AddonName clusters.AddonName = "loadimage"
+)
+
+type Addon struct {
+	image  string
+	loaded bool
+}
+
+func New() clusters.Addon {
+	return &Addon{}
+}
+
+// -----------------------------------------------------------------------------
+// CertManager Addon - Addon Implementation
+// -----------------------------------------------------------------------------
+
+func (a *Addon) Name() clusters.AddonName {
+	return AddonName
+}
+
+func (a *Addon) Deploy(ctx context.Context, cluster clusters.Cluster) error {
+	if a.image == "" {
+		a.loaded = true
+		return nil
+	}
+
+	switch ctype := cluster.Type(); ctype {
+	case kind.KindClusterType:
+		if err := a.loadIntoKind(cluster); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("loadimage addon is not supported by cluster type '%v'", cluster.Type())
+	}
+
+	return nil
+}
+
+func (a *Addon) Delete(ctx context.Context, cluster clusters.Cluster) error {
+	// per https://github.com/kubernetes-sigs/kind/issues/658 this is basically impossible
+	return fmt.Errorf("cannot remove loaded images from a cluster")
+}
+
+func (a *Addon) Ready(ctx context.Context, cluster clusters.Cluster) ([]runtime.Object, bool, error) {
+	// no way to verify this, we rely on Deploy's cmd.Run() not failing
+	return nil, a.loaded, nil
+}
```

Comment:
This is the first addon where `Deploy()` and `Ready()` are atomic: the others are all Deployments are similar where you get confirmation that the request to create succeeded, but readiness depends on an async process after that you must confirm separately.

Here, once `kind load docker-image` returns successfully, we're assured that the image is available (barring a hypothetical very odd bug in the command itself). That pattern should remain the same in any other implementations I can think of (absent a non-registry push like this, you're pushing to some registry, and `docker push` only succeeds if it actually pushed the image). My comment isn't correct--we actually can verify readiness, we just know it from earlier rather than needing to perform some additional check.

As a general pattern for atomic deploys, do you think this approach is sound? I think it's probably superfluous (AFAIK the `Deploy()` error will bubble up and test clients should abort then unless they ignore), but it otherwise still maintains the contract: `Ready()` does still report whether the addon is ready.

Interrogating containerd wouldn't be possible on registry-based solutions unless there's some way to list images on a remote registry (you won't be able to ssh into the kubelets to run crictl). That isn't a standard Docker feature (though a request has been open since 2015!), though it may be possible with individual registry providers through vendor-specific APIs, e.g. GCR [does provide this](https://cloud.google.com/container-registry/docs/managing#listing_images_by_their_storage_location).